### PR TITLE
RELENG-779 - Port sign-mozilla-vpn into iscript

### PR DESCRIPTION
**Pull request by hneiva**

### CODE REVIEW

**Comment by hneiva** [comment]
Waiting on QA to verify if signed package is OK.
I'll add the task.json to braindump repo so we have a baseline

File: iscript/src/iscript/script.py [hunk 1]
```diff
@@ -35,6 +36,9 @@ async def async_main(config, task):
     elif behavior == "mac_notarize":
         await notarize_behavior(config, task)
         return
+    elif behavior == "mac_notarize_vpn":
```
Comment (escapewindow) [inline iscript/src/iscript/script.py:39]: You may need to update [this line](https://github.com/mozilla-releng/scriptworker-scripts/blob/master/iscript/src/iscript/data/i_task_schema.json#L66) -- you probably already did on mac-v3-signing20. For that matter, we probably have to update the schema in general.

File: iscript/src/iscript/macvpn.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    update_keychain_search_path,
+)
+from iscript.util import get_sign_config
+
+log = logging.getLogger(__name__)
+
+# TODO: Move provisioning profiles from braindump
+LOGINITEM_PROVISIONING_PROFILE = "https://hg.mozilla.org/build/braindump/raw-file/tip/signing-related/firefoxvpn_loginitem_developerid.provisionprofile"
+NATIVEMESSAGING_PROVISIONING_PROFILE = "https://hg.mozilla.org/build/braindump/raw-file/tip/signing-related/FirefoxVPN_Native_Messaging.provisionprofile"
+MOZILLA_VPN_PROVISIONING_PROFILE = "https://hg.mozilla.org/build/braindump/raw-file/tip/signing-related/firefoxvpn_developerid.provisionprofile"
```
Comment (escapewindow) [inline iscript/src/iscript/macvpn.py]: 👍 if we expect these to change ever, we could plan to land these in the mozillavpn repo, and we could link to revision-pinned raw files, allowing devs to update these and ride the trains.

File: iscript/src/iscript/macvpn.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+async def _create_notarization_zipfile(work_dir, source, dest):
```
Comment (escapewindow) [inline iscript/src/iscript/macvpn.py:30]: This works. I thought about putting the `0` into the kwargs of `create_one_notarization_zipfile` but meh.

We can probably move this back into that function when we stop manually signing mozillavpn.

File: iscript/src/iscript/macvpn.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    assert len(top_app) == 1
+
+    await extract_all_apps(config, top_app)
+
+    # Extract fills in app.parent_dir
+    top_app = top_app[0]
+
+    ##############
+    # TODO: Decide what to do with the incoming package
+    # TODO: Remove when we switch to .tar.gz payloads (1/2)
```
Comment (escapewindow) [inline iscript/src/iscript/macvpn.py:100]: Yeah, we could determine this by task payload or by the original file's suffix or something; we may need to support both ways until we get this fully automated in mozillavpn taskgraph (or at least until we stop manually signing).

File: iscript/src/iscript/macvpn.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        artifact_prefix="public/",
+    )
+    await _sign_app(
+        config, task, sign_config, loginitems_app,
+        entitlements_url=task["payload"]["loginitems-entitlements-url"],
+        provisioning_profile_url=LOGINITEM_PROVISIONING_PROFILE
+    )
+
+    # NativeMessaging inner app
+    nativemessaging_app = App(
```
Comment (escapewindow) [inline iscript/src/iscript/macvpn.py:134]: Nice :)

File: iscript/src/iscript/macvpn.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    # zip_path = await create_one_notarization_zipfile(config["work_dir"], [app], sign_config, path_attrs=['app_path'])
+    zip_path = os.path.join(config["work_dir"], 'notarization.zip')
+    await _create_notarization_zipfile(config["work_dir"], top_app.pkg_path, zip_path)
+
+    poll_uuids = await notarize_no_sudo(config["work_dir"], sign_config, zip_path)
+    await poll_all_notarization_status(sign_config, poll_uuids)
+
+    log.info("Done notarizing app")
+
+    # TODO: Do we need to staple in place?
```
Comment (escapewindow) [inline iscript/src/iscript/macvpn.py]: I think there can be some warnings for the user if we don't staple and they launch it without network (or when the notarization service is down) -- best to staple if possible.

File: iscript/src/iscript/macvpn.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    # TODO: Decide what to do with the incoming package
+    # TODO: Remove when we switch to .tar.gz payloads (1/2)
+    build_zip_path = os.path.join(top_app.parent_dir, 'BUILD.zip')
+    await run_command([
+        "unzip", build_zip_path,
+        "-d", top_app.parent_dir
+    ])
+    os.remove(build_zip_path)
+
+    top_app.app_path = os.path.join(top_app.parent_dir, "Mozilla VPN.app")
```
Comment (escapewindow) [inline iscript/src/iscript/macvpn.py:105]: Noting that we're hardcoding signing a single app and single pkg.
This matches our current requirements. If the vpn team ever needs multiple apps/pkgs signed, then we may need to take an approach like `notarize_behavior` which has a list of `App` objects and signs each of them.

**Review by escapewindow — COMMENTED** [review]
Great! Outside of the schema change, which I've yet to see, looks like we're not going to negatively affect Firefox mac signing, while enabling MozillaVPN signing.

Did you modify the mac at all other than iscript? If yes, we need to doc those changes and plan on getting those into puppet someday. I'm thinking we haven't, though.

Let's get the schema change and start on enabling this on adhoc-signing. We can point mac-v3-signing20 at adhoc-signing and "test" in production :)

**Comment by escapewindow** [comment]
Looks like some linting errors as well, should be easy to fix :)

File: iscript/tests/test_macvpn.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    def get_sign_config(*args, **kwargs):
+        return {"signing_keychain": None, "keychain_password": None}
+
+    mocker.patch.object(macvpn, "_sign_app", new=noop_async)
+    mocker.patch.object(macvpn, "_create_notarization_zipfile", new=noop_async)
+    mocker.patch.object(macvpn, "get_app_paths", new=get_app_paths)
+    mocker.patch.object(macvpn, "extract_all_apps", new=noop_async)
+    mocker.patch.object(macvpn, "run_command", new=noop_async)
+    mocker.patch.object(macvpn.os, "remove", new=noop_sync)
+    mocker.patch.object(macvpn.os, "remove", new=noop_sync)
```
Comment (hneiva) [inline iscript/tests/test_macvpn.py]: oops

**Review by escapewindow — APPROVED** [review]
I think this is the current situation:

- you have mac-v3-signing20 quarantined for dev
- but haven't made any significant changes there other than iscript
- it may be worth verifying that [these steps](https://github.com/mozilla-releng/scriptworker-scripts/wiki/Testing-iscript#populate-the-work-dir-test) work properly, just to make sure we haven't broken anything with Firefox mac signing
- but yes, looks good, thank you! We can switch to the adhoc-signing task when you're ready.

**Pull request merged**